File: Itido-BMS/RentAPI/Rent/Models/TimePlanning/Absence.cs
using System;
using System.Linq.Expressions;

namespace Rent.Models.TimePlanning
{
    public class Absence
    {
        public int ID { get; set; }
        public int UserID { get; set; }
        public bool IsRequest { get; set; }
        public int CreatorID { get; set; }
        public int AbsenceReasonID { get; set; }
        public string Comment { get; set; }

        public DateTime From { get; set; }
        public DateTime To { get; set; }
        public int RequestID { get; set; }

        public virtual User User { get; set; }
        public virtual User Creator { get; set; }
        public virtual AbsenceReason AbsenceReason { get; set; }
        public virtual Request Request { get; set; }

        public static Expression<Func<Absence, dynamic>> BasicDTO(int requester, string requesterRole)
        {
            return a => a != null ?
            new
            {
                a.ID,
                a.IsRequest,
                a.From,
                a.To,
                AbsenceReason = AbsenceReason.BasicDTO().Compile()(a.AbsenceReason),
                Request = (a.Request != null && a.Request.Creator != null && a.Request.Subject != null) ? Request.BasicDTO(requester, requesterRole).Compile()(a.Request, a.Request.Creator) : null,
                User = User.BasicDTO().Compile()(a.User),
                Creator = User.BasicDTO().Compile()(a.Creator),
            } : null;
        }


        public static Expression<Func<Absence, dynamic>> StandardDTO(int requester, string requesterRole)
        {
            return a => a != null ?
            new
            {
                a.ID,
                a.IsRequest,
                a.From,
                a.To,
                AbsenceReason = AbsenceReason.BasicDTO().Compile()(a.AbsenceReason),
                Request = (a.Request != null && a.Request.Creator != null && a.Request.Subject != null) ? Request.BasicDTO(requester, requesterRole).Compile()(a.Request, a.Request.Creator) : null,
                User = User.BasicDTO().Compile()(a.User),
                Creator = User.BasicDTO().Compile()(a.Creator),
            } : null;
        }
    }
}

File: Itido-BMS/dart/bms_dart/lib/src/models/absence.dart
import 'package:bms_dart/src/models/absence_reason.dart';
import 'package:bms_dart/src/models/date_time_converter.dart';

import 'approval_state.dart';
import 'request.dart';
import 'user.dart';

class Absence {
  int id;
  String comment;
  //String description;
  DateTime from;
  DateTime to;
  bool isRequest;
  //bool canRespondToApprovalState;
  User creator;
  AbsenceReason absenceReason;
  Request request;

  Absence({
    this.id,
    this.comment,
    this.from,
    this.to,
    this.isRequest,
    //this.canRespondToApprovalState,
  });

  factory Absence.fromJson(json, {rootJson, path}) {
    if (json == null) return null;
    return Absence._fromJson(json, rootJson ?? {}, path ?? '');
  }

  Absence._fromJson(json, rootJson, path)
      : this.id = json['id'],
        // this.description = json['description'],
        this.comment = json['comment'],
        this.from = toDateTime(json['from']),
        this.to = toDateTime(json['to']),
        this.isRequest = json['isRequest'],
        //this.canRespondToApprovalState = json['canRespondToApprovalState'],
        this.creator = User.fromJson(json['creator']),
        this.absenceReason = AbsenceReason.fromJson(
            json['absenceReason'] ?? rootJson['${path}_absenceReason']),
        this.request = Request.fromJson(json['request']);

  Map<String, dynamic> toMap() => {
        'id': this.id,
        // 'description': this.description,
        'comment': this.comment,
        'from': this.from?.toString(),
        'to': this.to?.toString(),
        'isRequest': this.isRequest,
        //'canRespondToApprovalState': this.canRespondToApprovalState,
      };
}
